docs/design/variadics.md

```diff
@@ -0,0 +1,608 @@
+
+-   `...,` iteratively evaluates an expression, and treats the values as
+    successive elements of an enclosing tuple.
+-   `...and` and `...or` iteratively evaluate a boolean expression, combining
+    the values using `and` and `or`, and ending the loop early if the underlying
+    operator short-circuits.
+-   `...{}` iteratively executes a block of statements.
+
+`...,` can also be used in patterns, where it iteratively matches the elements
+of a scrutinee tuple. In conjunction with variadic bindings, this enables
+functions to take an arbitrary number of arguments.
+
+This example illustrates many of the key concepts:
+
+```carbon
+// Takes an arbitrary number of vectors with arbitrary element types, and
+// returns a vector of tuples where the i'th element of the vector is
+// a tuple of the i'th elements of the input vectors.
+fn Zip[each ElementType:! type]
+      (..., each vector: Vector(each ElementType))
+      -> Vector((..., each ElementType)) {
+  var iters: auto = (..., each vector.Begin());
+  var result: Vector((..., each ElementType));
+  while (...and expand iters != each vector.End()) {
+    result.push_back((..., *expand iters));
+    ...{ (expand iters)++; }
+  }
+  return result;
+}
+```
+
+`...,`, `...and`, and `...or` are prefix unary expression operators with the
+same precedence as `,`, `and`, and `or`, respectively. They are single tokens,
+and may not contain internal whitespace. `...{` is an opening delimiter that
+forms a balanced pair with `}`, and can contain any number of statements. An AST
+rooted at any of these operations is called a _pack expansion_. The operand of
+`...,`, `...and`, or `...or`, or the contents of the block delimited by
+`...{ }`, is called the _expansion body_.
+
+A pack expansion must contain one or more _expansion arguments_, which are
+marked by `expand` or `each`. `expand` is a prefix unary expression operator
+with the same precedence as `*`. The operand of `each` is always an identifier
+name or a binding pattern, so it does not have a precedence. For example, in the
+loop condition `...and expand iters != each vector.End()` in the body of `Zip`,
+`expand iters` and `each vector` are expansion arguments. Note that `each` can
+also appear in a deduced parameter list, as discussed below.
+
+The _arity_ of an expansion argument is a compile-time value representing the
+number of elements it evaluates to (or the number of values it matches, if it's
+a pattern). Every pack expansion must contain at least one expansion argument,
+and all arguments of a given expansion must have the same arity (which we will
+also refer to as the arity of the expansion). If any expansion argument is a
+non-expression pattern, the entire pack expansion is a pattern. In particular,
+this means that the expansion arguments of a `...{` expansion must be
+expressions.
+
+Pack expansions can be nested only if the inner expansion is within one of the
+outer expansion's arguments. For example,
```

Comment:
It's not clear to me what this means -- what are the arguments of the outer expansion?

Is the idea here that you can't nest a pack expansion in another pack expansion "at the top level"? Eg, `(..., ..., expand each tuple)` wouldn't be permitted, but `(..., (..., expand each tuple))` would be. Or that you can't form a pack-of-packs, so `expand each tuple` would itself be invalid?

Reply:
Oh, perhaps this means "within the outer expansion's <ins>expansion</ins> arguments"? Which I think would translate into a "no packs of packs" restriction.